### PR TITLE
[NN] Add a learned laplacian positional encoder

### LINKED ISSUE #4746: Add laplacian positional encoder
## 🔨Work Item
Add a learned laplacian positional encoder for encoding eigenvalues and eigenvectors.

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by ZHITENGLI**
## Description

close #4746 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes

Add a learned laplacian positional encoder for encoding eigenvalues and eigenvectors.

### CODE REVIEW

**Comment by ZHITENGLI** [comment]
@rudongyu

File: python/dgl/nn/pytorch/utils.py [hunk 1]
```diff
@@ -555,3 +555,154 @@ def forward(self, g, labels, mask=None):
                     y[mask] = labels[mask]
 
             return y
+
+
+class LaplacianPosEnc(nn.Module):
+    r"""Laplacian Positional Encoder, as introduced in
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/utils.py:561]:
```suggestion
    r"""Laplacian Positional Encoder (LPE), as introduced in
```

File: python/dgl/nn/pytorch/utils.py [hunk 2]
```diff
@@ -555,3 +555,154 @@ def forward(self, g, labels, mask=None):
...
+    `GraphGPS: General Powerful Scalable Graph Transformers
+    <https://arxiv.org/abs/2205.12454>`__
+    This module is a learned laplacian positional encoding module using Transformer or DeepSet.
+
+    Parameters
+    ----------
+    model_type : str
+        Encoder NN model type for LPE, can only be "Transformer" or "DeepSet".
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/utils.py:569]:
```suggestion
        Encoder model type for LPE, can only be "Transformer" or "DeepSet".
```

File: python/dgl/nn/pytorch/utils.py [hunk 3]
```diff
@@ -555,3 +555,154 @@ def forward(self, g, labels, mask=None):
...
+    num_layer : int
+        Number of layers in Transformer/DeepSet Encoder.
+    max_freqs : int
+        Number of eigenvectors.
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/utils.py:573]:
```suggestion
    k : int
        Number of smallest non-trivial eigenvectors.
```
Unify the notation of this parameter with `LaplacianPE` as `k` may be better. Please modify the notation in other places accordingly.

File: python/dgl/nn/pytorch/utils.py [hunk 4]
```diff
@@ -555,3 +555,154 @@ def forward(self, g, labels, mask=None):
...
+    lpe_dim : int
+        Output size of final laplacian encoding.
+    n_head : int, optional
+        Number of heads in Transformer Encoder.
+        Default : 1.
+    batch_norm : bool, optional
+        If True, implement batch normalization on raw LaplacianPE.
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/utils.py:580]:
```suggestion
        If True, apply batch normalization on raw LaplacianPE.
```

File: python/dgl/nn/pytorch/utils.py [hunk 5]
```diff
@@ -555,3 +555,154 @@ def forward(self, g, labels, mask=None):
...
+        Default : False.
+    post_n_layer : int, optional
+        If post_n_layer > 0, apply an MLP after pooling.
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/utils.py:583]:
```suggestion
    num_post_layer : int, optional
        If num_post_layer > 0, apply an MLP of ``num_post_layer`` layers after pooling.
```

File: python/dgl/nn/pytorch/utils.py [hunk 6]
```diff
@@ -555,3 +555,154 @@ def forward(self, g, labels, mask=None):
...
+        num_layer,
+        max_freqs,
+        lpe_dim,
+        n_head=1,
+        batch_norm=False,
+        post_n_layer=0,
+    ):
+        super(LaplacianPosEnc, self).__init__()
+        self.model_type = model_type
+        self.linear_A = nn.Linear(2, lpe_dim)
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/utils.py:616]: maybe use `self.linear` as the attribute name, following the convention.

File: python/dgl/nn/pytorch/utils.py [hunk 7]
```diff
@@ -555,3 +555,154 @@ def forward(self, g, labels, mask=None):
...
+            self.post_mlp = nn.Sequential(*layers)
+        else:
+            self.post_mlp = None
+
+    def forward(self, EigVals, EigVecs):
+        r"""
+        Parameters
+        ----------
+        EigVals : Tensor
+            Laplacian Eigenvalues of shape :math:`(N, k)`, can be obtained by using `LaplacianPE`.
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/utils.py:671]: please elaborate that the eigenvalues are repeated on the first dimension.

File: tests/pytorch/test_nn.py [hunk 8]
```diff
@@ -1666,3 +1666,22 @@ def test_MetaPath2Vec(idtype):
     model = model.to(dev)
     embeds = model.node_embed.weight
     assert embeds.shape[0] == g.num_nodes()
+
+def test_LaplacianPosEnc():
```
Comment (rudongyu) [inline tests/pytorch/test_nn.py:1670]: you may use `@pytest.mark.parametrize` to test multiple groups (e.g. w or w/o batch_norm, different num_layer, different num_post_layer, etc.)

**Review by rudongyu — COMMENTED** [review]
finished the 1st pass.

**Comment by rudongyu** [comment]
@dgl-bot

**Comment by rudongyu** [comment]
@dgl-bot

**Review by rudongyu — APPROVED** [review]
LGTM

**Pull request merged**